### DISCUSSION TITLE
Why P assumes send are reliable?

### DISCUSSION THREAD

**Discussion by gabrielgiussi** [p-org/P#669] [Q&A] [answered]
In [P semantics](https://p-org.github.io/P/advanced/psemantics/) says

> Sends are reliable, buffered, non-blocking, and directed (not broadcast). Sends are reliable i.e., executing a send operation in P adds an event into the target machines buffer. Hence, if one wants to model message loss it has to be modeled explicitly (discussed in the Failure Detector example in the tutorial).

Why is it safe to assume such semantics to develop a distributed system over an unreliable network?
Is it because we consider the TCP protocol already provides such guarantees? 

Isn't this leaving outside (although I understand the possibility to implement such a network abstraction exists) an important characteristic of distributed systems that is the impossibility to distinguish a slow node from a crashed one?

**Comment by AnDongLi** [comment, marked as answer]
you can model unreliable send like this: (example from tutorial)
```
fun UnReliableSend(target: machine, message: event, payload: any) {
  // nondeterministically drop messages
  // $: choose()
  if($) send target, message, payload;
}
```
So send itself can be considered always reliable.
